Content/Items/Haunted/Weapons.EchochainWhip.cs

```diff
@@ -0,0 +1,703 @@
+using Terraria.Enums;
+using Terraria.GameContent.Creative;
+using Terraria.Graphics.Effects;
+using Terraria.ID;
+
+namespace StarlightRiver.Content.Items.Haunted
+{
+	public class EchochainWhip : ModItem
+	{
+		internal int cooldown;
+		public override string Texture => AssetDirectory.HauntedItem + Name;
+
+		public override void SetStaticDefaults()
+		{
+			DisplayName.SetDefault("Echochain");
+			Tooltip.SetDefault("Chains enemies together, sharing summon damage between them\nHold and release <right> to snare enemies near your mouse to the ground, chaining all effected enemies");
+			CreativeItemSacrificesCatalog.Instance.SacrificeCountNeededByItemId[Type] = 1;
+		}
+
+		public override void SetDefaults()
+		{
+			Item.DefaultToWhip(ModContent.ProjectileType<EchochainWhipProjectile>(), 15, 3f, 3.75f, 40);
+			Item.SetShopValues(ItemRarityColor.Green2, Item.sellPrice(gold: 1));
+		}
+
+		public override bool AltFunctionUse(Player player)
+		{
+			return cooldown <= 0;
+		}
+
+		public override void UpdateInventory(Player player)
+		{
+			if (cooldown > 0)
+				cooldown--;
+		}
+
+		public override bool CanUseItem(Player player)
+		{
+			if (player.altFunctionUse == 2)
+			{
+				Item.UseSound = SoundID.DD2_BetsyFireballImpact;
+				Item.useStyle = ItemUseStyleID.Shoot;
+			}
+			else
+			{
+				Item.UseSound = SoundID.Item152;
+				Item.useStyle = ItemUseStyleID.Swing;
+			}
+
+			return base.CanUseItem(player);
+		}
+
+		public override bool Shoot(Player player, EntitySource_ItemUse_WithAmmo source, Vector2 position, Vector2 velocity, int type, int damage, float knockback)
+		{
+			if (player.altFunctionUse == 2 && cooldown <= 0)
+			{
+				Projectile.NewProjectile(source, position, velocity, ModContent.ProjectileType<EchochainWhipAltProjectile>(), damage, knockback, player.whoAmI);
+				cooldown = 120;
+				return false;
+			}
+
+			return true;
+		}
+
+		public override void AddRecipes()
+		{
+			Recipe recipe = CreateRecipe();
+
+			recipe.AddIngredient(ItemID.GoldBar, 15);
+			recipe.AddIngredient<VengefulSpirit>(10);
+			recipe.AddTile(TileID.Anvils);
+			recipe.Register();
+
+			recipe = CreateRecipe();
+
+			recipe.AddIngredient(ItemID.PlatinumBar, 15);
+			recipe.AddIngredient<VengefulSpirit>(10);
+			recipe.AddTile(TileID.Anvils);
+			recipe.Register();
+		}
+	}
+
+	public class EchochainWhipAltProjectile : ModProjectile
+	{
+		public bool playedSound;
+
+		public float handRotation;
+		public Vector2 oldMouse;
+
+		public Point16?[] tiles = new Point16?[18]; // tiles which will glow visually
+		public Point16?[] enemyTiles = new Point16?[6]; // tiles which enemies will be chained onto
+		public NPC[] targets = new NPC[6];
+
+		public ref float DeathTimer => ref Projectile.ai[0];
+
+		public Vector2 OwnerMouse => Owner.GetModPlayer<ControlsPlayer>().mouseWorld;
+		public Player Owner => Main.player[Projectile.owner];
+
+		public bool CanHold => Owner.GetModPlayer<ControlsPlayer>().mouseRight && !Owner.CCed && !Owner.noItems;
```

Comment:
you need to enable rightClickListener on this controls player every frame this is needed so it can be sent to the other clients